File: iso27001-gin/internal/core/middleware/security_headers.go
// Package middleware contains Gin middleware for security, telemetry, and rate limiting.
package middleware

import (
	"github.com/gin-gonic/gin"
)

// SecurityHeaders injects A.10-mandated HTTP security headers on every response.
// Mirrors FastAPI SecurityHeadersMiddleware and Symfony SecurityHeaderSubscriber.
func SecurityHeaders() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Header("Strict-Transport-Security", "max-age=31536000; includeSubDomains")
		c.Header("X-Frame-Options", "DENY")
		c.Header("X-Content-Type-Options", "nosniff")
		c.Header("Referrer-Policy", "strict-origin-when-cross-origin")
		c.Header("Content-Security-Policy", "default-src 'none'; frame-ancestors 'none'")
		c.Header("Permissions-Policy", "geolocation=(), microphone=(), camera=()")
		c.Header("Cross-Origin-Opener-Policy", "same-origin")
		c.Header("Cross-Origin-Embedder-Policy", "require-corp")
		// Remove server fingerprint headers
		c.Header("Server", "")
		c.Header("X-Powered-By", "")
		c.Next()
	}
}

File: iso27001-phoenix/lib/iso27001_phoenix/core/middleware/security_headers.ex
defmodule Iso27001Phoenix.Core.Middleware.SecurityHeaders do
  @moduledoc """
  A.10: Plug that adds all required security headers and removes
  server-identifying headers on every response.
  """
  import Plug.Conn

  @behaviour Plug

  @impl true
  def init(opts), do: opts

  @impl true
  def call(conn, _opts) do
    conn
    |> delete_resp_header("server")
    |> delete_resp_header("x-powered-by")
    |> put_resp_header("strict-transport-security", "max-age=31536000; includeSubDomains; preload")
    |> put_resp_header("x-frame-options", "DENY")
    |> put_resp_header("x-content-type-options", "nosniff")
    |> put_resp_header("content-security-policy", "default-src 'none'")
    |> put_resp_header("x-xss-protection", "1; mode=block")
    |> put_resp_header("referrer-policy", "strict-origin-when-cross-origin")
    |> put_resp_header("permissions-policy", "geolocation=(), microphone=(), camera=()")
    |> put_resp_header("cache-control", "no-store")
  end
end
